implementation/src/main/java/io/smallrye/graphql/execution/QueryCache.java

```diff
@@ -0,0 +1,98 @@
+
+import org.jboss.logging.Logger;
+
+import graphql.ExecutionInput;
+import graphql.execution.instrumentation.InstrumentationContext;
+import graphql.execution.instrumentation.SimpleInstrumentation;
+import graphql.execution.instrumentation.parameters.InstrumentationValidationParameters;
+import graphql.execution.preparsed.PreparsedDocumentEntry;
+import graphql.execution.preparsed.PreparsedDocumentProvider;
+import graphql.validation.ValidationError;
+
+public class QueryCache extends SimpleInstrumentation implements PreparsedDocumentProvider {
+    private static final Logger LOG = Logger.getLogger(QueryCache.class);
+    private static final int MAX_CACHE_SIZE = AccessController.doPrivileged((PrivilegedAction<Integer>) () -> {
+        return Integer.getInteger("io.smallrye.graphql.execution.queryCacheMaxSize", 2048);
+    });
+
+    private final LRUCache<String, PreparsedDocumentEntry> cache = new LRUCache<>(MAX_CACHE_SIZE);
+    private final ThreadLocal<ExecutionFunction> executionFunctionTL = new ThreadLocal<>();
+
+    @Override
+    public PreparsedDocumentEntry getDocument(ExecutionInput executionInput,
+            Function<ExecutionInput, PreparsedDocumentEntry> computeFunction) {
+        String query = executionInput.getQuery();
+        PreparsedDocumentEntry entry = cache.get(query);
+        if (entry == null) {
+            ExecutionFunction executionFunction = new ExecutionFunction(computeFunction, executionInput);
+            executionFunctionTL.set(executionFunction);
+            entry = computeFunction.apply(executionInput);
+        } else if (LOG.isDebugEnabled()) {
+            LOG.debug("Retrieved from cache: " + query);
+        }
+        return entry;
+    }
+
+    @Override
+    public InstrumentationContext<List<ValidationError>> beginValidation(
+            InstrumentationValidationParameters parameters) {
+
+        ExecutionFunction executionFunction = executionFunctionTL.get();
+        if (executionFunction != null) {
+            executionFunctionTL.remove();
+            return new ValidationInstrumentationContext(executionFunction);
+        }
+        return super.beginValidation(parameters);
+    }
+
+    private static class ExecutionFunction implements Function<String, PreparsedDocumentEntry> {
+        private final Function<ExecutionInput, PreparsedDocumentEntry> function;
+        private final ExecutionInput executionInput;
+
+        ExecutionFunction(Function<ExecutionInput, PreparsedDocumentEntry> function, ExecutionInput executionInput) {
+            this.function = function;
+            this.executionInput = executionInput;
+        }
+
+        @Override
+        public PreparsedDocumentEntry apply(String s) {
+            return function.apply(executionInput);
+        }
+
+        String getQuery() {
+            return executionInput.getQuery();
+        }
+    }
+
+    private class ValidationInstrumentationContext implements InstrumentationContext<List<ValidationError>> {
```

Comment:
If you made the cache static from the comment above, you could make this inner class static as well.